### ISSUE TITLE
Dangerous use of ForkJoinPool

### ISSUE THREAD

**Issue by twothe** [corona-warn-app/cwa-server#399]
https://github.com/corona-warn-app/cwa-server/blob/822150b1b22645ba071a6f6576d50614dc01b34b/services/submission/src/main/java/app/coronawarn/server/services/submission/controller/SubmissionController.java#L63

The common ForkJoinPool is used to process all real incoming transmissions. This will cause the pool to be flooded under heavy load and block all other threaded streaming operations, which could potentially put the whole application into a dead-lock. 

Also no forking and joining is done in the processing, which means the wrong type is used here. You should to use a standard Executor instead.

In general the DeferredResult should only be used if there is no alternative, as it uses dangerous locking, unnecessary synchronization and the [double-checked locking anti-pattern](https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java).

**Comment by nils4cosee** [comment]
The `ForkJoinPool.commonPool()` has by default the size of CPUs minus 1 threads, and I think it is meant to be used for CPU-intensive tasks.

In this case, the controller calls the verification-service and the database, which should be the most time-consuming parts of its job. I think Webflux was made for exactly this kind of problem (but I haven't used it yet).

Alternatively, I would use custom ThreadPoolExecutor with a higher thread-count.

**Comment by twothe** [comment]
The [ForkJoinPool](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html) exists to allow external work-stealing algorithms to use an easily available Executor. It is **not** meant to be used as a general all-purpose Executor and issues can arise if it is used in such a way, as those resources are then no longer available for their dedicated intent. 

> The pool attempts to maintain enough active (or available) threads [..] even if some tasks are stalled waiting to join others. However, no such adjustments are guaranteed in the face of blocked I/O or other unmanaged synchronization.

In other words: if you just randomly throw in tasks into that pool it might get stuck. This is especially dangerous when used in the main loop of a web server, because then any flooding of requests also immediately leads to a flooding of the one single common pool, which is like rolling out a red carpet for hackers.

As the appropriate alternatives can be used very easily, and it takes just a very minor adjustment to the code, there is no reason to work with the wrong type of Executor.

In general a web-server thread should not open another thread unless multiple long-time blocking operations are performed simultaneously. There is no performance gain associated with that: if all cores are busy handling requests opening yet another thread will not cause the server to run faster, but in fact slower. It however makes the code more complicated and more error-prone (as beautifully demonstrated by the code in question). Therefore using DeferredResult or threading in general in this particular context is a completely unnecessarily complex implementation for a very simple task. The KISS principle should be applied instead.

**Comment by nils4cosee** [comment]
I agree that the common ForkJoinPool is not adequate here (did I make that clear in my comment?). And I have also wondered if asynchronous processing makes sense at all.

As far as I can see, the asynchronous processing has been introduced when the fake-requests were added. (see #109)

1) For handling the fake requests, asynchronous processing makes much sense, because there is no actual computation, just a delay before the request in answered.
2) For handling real requests, we can probably do the computation in the same thread, because it is the same controller, we still need to return a DeferredResult.

I think there will be many more fake-requests than real requests, so we should optimize for those.

Or we build everything with Flux. I am currently doing some performance tests.

@stevesap I couldn't find any infrastructure-as-a-service repository so far. Do you know how instances will be scaled in the cloud?

**Comment by stevesap** [comment]
@nils4cosee thanks for your efforts there. I am working on getting an answer to this ..

**Comment by nils4cosee** [comment]
I have used a fresh spring-boot-project to do some performance tests with the following controller:

```java
@RestController
public class HelloController {


  private static ScheduledExecutorService executorService = Executors.newScheduledThreadPool(10);

  @RequestMapping("/deferredResult")
  public DeferredResult<String> index() throws InterruptedException {
    DeferredResult<String> stringDeferredResult = new DeferredResult<>();

    executorService.schedule(() -> stringDeferredResult.setResult("hello"),
        1000, TimeUnit.MILLISECONDS);
    return stringDeferredResult;
  }

  @RequestMapping("/flux")
  public Mono<String> flux() throws InterruptedException {
    return Mono.delay(Duration.of(
        1000,
        ChronoUnit.MILLIS
    )).then(Mono.just("hello"));
  }
}
```

I have used JMeter to fire 100 requests in each of  1000 parallel threads to each endpoint and both performed equally well.

I think the solution could be something like fd76c872a37151706520c2774245428c59b9e24d:

Use an executor for fake-requests. Do not use other threads for persisting actual results.

**Comment by stevesap** [comment]
@nils4cosee - we are using standard Kubernetes capabilities to scale our application, and our performance testers are continuously fine tuning the parameters to different load/attack scenarios.

**Comment by twothe** [comment]
That looks like a good solution.

While you are at performance testing: it might be a good idea to compare your implementation to a primitive implementation using only `Thread.sleep` and no threading or anything fancy. That will show the benefit of your implementation over it in clear numbers and might provide some insight on how to configure the servers later on.

**Comment by sameer-dighe** [comment]
Hi @twothe, I came across this discussion and my problem is kind of same. I am using below implementation for fork join which parallel the execution of each message and it does that but in reality the performance is not getting boost when I increase the instances and try to increase a performance. At a time 10 messages are polled from queue to start the parallel execution. can you help me to understand whether I did something wrong in code.here is my code.

```
public class StockRangeConsumer {

    private static final Logger LOGGER = LoggerFactory.getLogger(StockRangeConsumer.class);

    public static final int THREADS = 100;

    private final ObjectMapper objectMapper;
    private final SqsMessageConsumer stockRangeConsumer;
    private final ForkJoinPool forkJoinPool = new ForkJoinPool(THREADS);

    private final StockRangeDataProcessor stockRangeDataProcessor;
    private final PilotFacilityFilterService pilotFacilityFilterService;

    public StockRangeConsumer(StockRangeDataProcessor stockRangeDataProcessor,
                              @Qualifier(StockRangeConfiguration.STOCK_RANGE_CONSUMER) SqsMessageConsumer stockRangeConsumer,
                              ObjectMapper objectMapper,
                              PilotFacilityFilterService pilotFacilityFilterService) {
        this.pilotFacilityFilterService = pilotFacilityFilterService;
        this.objectMapper = objectMapper;
        this.stockRangeConsumer = stockRangeConsumer;
        this.stockRangeDataProcessor = stockRangeDataProcessor;
    }

    public void consume() throws ExecutionException, InterruptedException {
        List<Message> messages = stockRangeConsumer.retrieve();
        LOGGER.info("Number of available core in the processor is: {}", Runtime.getRuntime().availableProcessors());

        forkJoinPool.submit(() -> messages.parallelStream().forEach(message -> {
            try {
                RangePredictionData rangePredictionData = toRangePrediction(message.getMessage());
                LOGGER.info("Starting run analysis for part key: {}", rangePredictionData.getPartKey());
                LOGGER.debug("***** RangePredictionData with PartKey: {}, list size: {}", rangePredictionData.getPartKey(), messages.size());
                processRangePredictionData(rangePredictionData);
                LOGGER.info("Ended run analysis for part key: {}", rangePredictionData.getPartKey());
                stockRangeConsumer.delete(message);
            } catch (Exception e) {
                LOGGER.error("Error reading message from stock range queue : ", e);
            }
        })).get();

    }

    // TODO move logic to usecase/service and this class to adapter
    public void processRangePredictionData(RangePredictionData rangePredictionData) {
        if (rangePredictionData == null || rangePredictionData.getPartKey() == null || rangePredictionData.getRangeCalculationData() == null) {
            LOGGER.error("Range prediction consumed, but is null or missing range calculation data");
            return;
        }

        if (pilotFacilityFilterService.isPilotFacilityAndExcludedUsageCode(rangePredictionData.getPartKey().getFacilityCode(), rangePredictionData.getPartKey().getUsageCode())) {
            stockRangeDataProcessor.processRangePredictionData(rangePredictionData);
        }
    }

    private RangePredictionData toRangePrediction(String message) throws JsonProcessingException {
        return objectMapper.readValue(message, RangePredictionData.class);
    }

}
```
